include/irr/builtin/material_compiler/glsl/common.glsl

```diff
@@ -0,0 +1,1386 @@
+		uint op = instr_getOpcode(instr);
+
+		if (INSTR_1ST_DWORD(instr) != INSTR_1ST_DWORD(generator))
+		{
+			instr_eval_and_pdf_execute(instr, s, uf, ts);
+		}
+		else
+		{
+			//assert(isBXDF(instr));
+			uvec3 regs = instr_decodeRegisters(instr);
+			writeReg(REG_DST(regs), eval_and_pdf_t(0.0));
+		}
+
+#if defined(OP_SET_GEOM_NORMAL)||defined(OP_BUMPMAP)
+		if (
+#ifdef OP_SET_GEOM_NORMAL
+			op == OP_SET_GEOM_NORMAL
+#ifdef OP_BUMPMAP
+			||
+#endif
+#endif
+#ifdef OP_BUMPMAP
+			op == OP_BUMPMAP
+#endif
+			) {
+			s = irr_glsl_createLightSample(s.L, currInteraction);
+			uf = irr_glsl_calcAnisotropicMicrofacetCache(currInteraction, s);
+		}
+#endif
+	}
+
+	eval_and_pdf_t eval_and_pdf = readReg4(0u);
+	return eval_and_pdf;
+}
+
+irr_glsl_AnisotropicMicrofacetCache getSmoothMicrofacetCache(in float NdotV, in float NdotL)
+{
+	irr_glsl_AnisotropicMicrofacetCache uf;
+	const float d = 1e-4;//some delta to avoid zeroes... not a perfect solution but for now..
+	uf.isotropic.VdotH = NdotV-d;
+	uf.isotropic.LdotH = NdotL-d;
+	uf.isotropic.NdotH = d;
+	uf.isotropic.NdotH2 = d*d;
+	uf.TdotH = 1.0-d;
+	uf.BdotH = 1.0-d;
+
+	return uf;
+}
+
+irr_glsl_LightSample irr_bsdf_cos_generate(in instr_stream_t stream, in vec3 rand, out vec3 out_remainder, out float out_pdf, out instr_t out_generatorInstr, out irr_glsl_AnisotropicMicrofacetCache out_uf)
+{
+	uint ix = 0u;
+	instr_t instr = irr_glsl_MC_fetchInstr(stream.offset);
+	uint op = instr_getOpcode(instr);
+	float weight = 1.0;
+	vec3 u = rand;
+
+	instr_execute_SET_GEOM_NORMAL();
+	while (!op_isBXDF(op))
+	{
+#ifdef OP_BLEND
+		if (op==OP_BLEND) {
+			bsdf_data_t bsdf_data = fetchBSDFDataForInstr(instr);
+			params_t params = instr_getParameters(instr, bsdf_data);
+			float w = params_getBlendWeight(params);
+			float rcpChoiceProb;
+			bool choseRight = irr_glsl_partitionRandVariable(w, u.z, rcpChoiceProb);
+
+			uint right = instr_getRightJump(instr);
+			ix = choseRight ? right:(ix+1u);
+			weight *= choseRight ? (1.0-w):w;
+		}
+		else 
+#endif //OP_BLEND
+		{
+#ifdef OP_SET_GEOM_NORMAL
+			if (op==OP_SET_GEOM_NORMAL)
+			{
+				instr_execute_SET_GEOM_NORMAL();
+			} else 
+#endif //OP_SET_GEOM_NORMAL
+#ifdef OP_BUMPMAP
+			if (op==OP_BUMPMAP)
+			{
+				mat2x4 srcs = instr_fetchSrcRegs(instr);
+				instr_execute_BUMPMAP(instr, srcs);
+			} else
+#endif //OP_BUMPMAP
+			{}
+			ix = ix + 1u;
+		}
+
+		instr = irr_glsl_MC_fetchInstr(stream.offset+ix);
+		op = instr_getOpcode(instr);
+	}
+
+	bsdf_data_t bsdf_data = fetchBSDFDataForInstr(instr);
+	params_t params = instr_getParameters(instr, bsdf_data);
+	//speculatively
+	const float ax = params_getAlpha(params);
+	const float ax2 = ax*ax;
+	const float ay = params_getAlphaV(params);
+	const float ay2 = ay*ay;
+	const mat2x3 ior = bsdf_data_decodeIoR(bsdf_data,op);
+	const bool is_bsdf = !op_isBRDF(op);
+	const vec3 refl = params_getReflectance(params);
+
+#ifndef NO_TWOSIDED
+	bool ts_flag = false;
+	handleTwosided_interactionOnly(ts_flag, instr);
```

Comment:
this is weird, it should just be
```glsl
if (instr_getTwosided(instr) && currInteraction.isotropic.NdotV<0.0)
{
	flipCurrInteraction();
}
```

Reply:
eh, you're right. I was thinking that it could get be reflipped in next instruction, but... well yeah. obvs not possible

Reply:
its the backend that should be propagating the twosided thing down the AST branches

Reply:
> its the backend that should be propagating the twosided thing down the AST branches

i dont get how this is relevant to those comments, however... thats how it is basically. IR has twosided (2 or 1 surface) in root node and backend is propagating the flag while generating instructions for tree traversal